Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp

```diff
@@ -276,12 +277,70 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
     // explicitly allow the following allocation(s).
     DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
 
+    if (!m_currentAudioSamplingRate) {
+        Locker locker { m_ringBufferLock };
+        m_ringBuffer = InProcessCARingBuffer::allocate(*std::get<const AudioStreamBasicDescription*>(description.platformDescription().description), description.sampleRate() * 2); // Allow for 2s of audio.
+        queueSingleton().dispatch([weakThis = ThreadSafeWeakPtr { *this }, description = *std::get<const AudioStreamBasicDescription*>(description.platformDescription().description)] {
+            if (RefPtr protectedThis = weakThis.get())
+                protectedThis->audioSamplesAvailableStarted(description);
+        });
+    }
     ASSERT(!m_currentAudioSamplingRate || m_currentAudioSamplingRate == description.sampleRate(), "sampling rate should never change over a session");
     m_currentAudioSamplingRate = description.sampleRate();
     m_lastEnqueuedAudioSampleCount = m_currentAudioSampleCount.load();
     m_currentAudioSampleCount += sampleCount;
-    if (RetainPtr sampleBuffer = createAudioSampleBuffer(data, description, PAL::CMTimeMake(m_lastEnqueuedAudioSampleCount, description.sampleRate()), sampleCount))
-        audioCompressor()->addSampleBuffer(sampleBuffer.get());
+    {
+        if (!m_ringBufferLock.tryLock())
+            return;
+        Locker locker { AdoptLock, m_ringBufferLock };
+        if (!m_ringBuffer)
+            return;
+        m_ringBuffer->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_lastEnqueuedAudioSampleCount);
+    }
+    queueSingleton().dispatch([weakThis = ThreadSafeWeakPtr { *this }, sampleCount, lastEnqueued = m_lastEnqueuedAudioSampleCount.load()] {
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->audioSamplesAvailable(sampleCount, lastEnqueued);
+    });
+}
+
+void MediaRecorderPrivateEncoder::audioSamplesAvailableStarted(const AudioStreamBasicDescription& description)
+{
+    assertIsCurrent(queueSingleton());
+
+    ASSERT(!m_audioFormatDescription);
+    CMFormatDescriptionRef newFormat = nullptr;
+    if (auto error = PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &description, 0, nullptr, 0, nullptr, nullptr, &newFormat)) {
+        RELEASE_LOG_ERROR(MediaStream, "MediaRecorderPrivateEncoder::audioSamplesAvailable: CMAudioFormatDescriptionCreate failed with %u", error);
+        return;
+    }
+    m_audioFormatDescription = adoptCF(newFormat);
+}
+
+void MediaRecorderPrivateEncoder::audioSamplesAvailable(size_t sampleCount, size_t totalSampleCount)
+{
+    assertIsCurrent(queueSingleton());
+
+    ASSERT(m_audioFormatDescription);
+
+    auto* absd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
+    ASSERT(absd);
```

Comment:
Why not log and return early if the absd is null, as you do for all of the other error conditions?

Reply:
If it "can't happen", why bother asserting? If it can happen, the we will crash in the dereference in a release build.